lib/inferno/dsl/auth_info.rb

```diff
@@ -168,6 +168,143 @@ def add_to_client(client)
 
         client.set_bearer_token(access_token)
       end
+
+      # @private
+      def need_to_refresh?
+        return false if access_token.blank? || (auth_type != 'backend_services' && refresh_token.blank?)
+
+        return true if expires_in.blank?
+
+        issue_time.to_i + expires_in.to_i - DateTime.now.to_i < 60
+      end
+
+      # @private
+      def able_to_refresh?
+        (auth_type == 'backend_services' && token_url.present?) || (refresh_token.present? && token_url.present?)
+      end
+
+      # @private
+      def oauth2_refresh_params
+        case auth_type
+        when 'public'
+          public_auth_refresh_params
+        when 'symmetric'
+          symmetric_auth_refresh_params
+        when 'asymmetric'
+          asymmetric_auth_refresh_params
+        when 'backend_services'
+          backend_services_auth_refresh_params
+        end
+      end
+
+      # @private
+      def symmetric_auth_refresh_params
+        {
+          'grant_type' => 'refresh_token',
+          'refresh_token' => refresh_token
+        }
+      end
+
+      # @private
+      def public_auth_refresh_params
+        symmetric_auth_refresh_params.merge('client_id' => client_id)
+      end
+
+      # @private
+      def asymmetric_auth_refresh_params
+        symmetric_auth_refresh_params.merge(
+          'client_assertion_type' => 'urn:ietf:params:oauth:client-assertion-type:jwt-bearer',
+          'client_assertion' => signed_jwt
+        )
+      end
+
+      # @private
+      def backend_services_auth_refresh_params
+        {
+          'grant_type' => 'client_credentials',
+          'scope' => requested_scopes,
+          'client_assertion_type' => 'urn:ietf:params:oauth:client-assertion-type:jwt-bearer',
+          'client_assertion' => signed_jwt
+        }
+      end
+
+      # @private
+      def oauth2_refresh_headers
+        base_headers = { 'Content-Type' => 'application/x-www-form-urlencoded' }
+
+        return base_headers unless auth_type == 'symmetric'
+
+        credentials = "#{client_id}:#{client_secret}"
+
+        base_headers.merge(
+          'Authorization' => "Basic #{Base64.strict_encode64(credentials)}"
+        )
+      end
+
+      # @private
+      def jwk_set
+        jwks_path = File.join(__dir__, 'jwks.json')
```

Comment:
Like we do for SMART App Launch and Bulk Data, we need to be able to use an environment variable to set the JWKS location, and use this as the default when that environment variable isn't present. `jwks.rb` from the SMART App Launch Test Kit could probably be brought in more or less as is to handle the jwks.